### PR TITLE
MM-51095: Foundation for ESR upgrade scripts

### PR DESCRIPTION
**Pull request by agarciamontoro**
#### Summary
This is still a draft PR, but I wanted to ask for feedback early on so we can keep iterating on it. Let me give an overview of the changes before you go and review it:
- There's a new Github action, ESR Upgrade, which intends to check all migration scripts we define against the migration done through the server. It does this by running both migrations on the same DB, dumping the result in each case and `diff`ing them. For now, the action only uploads the dumps as a Github action artifact, letting us manually inspect them while we iterate on both the action and the script. I want to make that automatic before we merge this PR.
- The migration script itself is, for now, a simple concatenation of all migrations between 54 and 101, both included.
- Running the migration script through the `mysql` CLI is not possible, since we **need** to let MySQL know that the file contains multiple statements (due to the stored procedures), and that's only possible through the `multiStatements=true` property in the DSN. I'd *love* to stand corrected on that, since this bit adds quite a lot of complexity; namely:
  - As we need to specify the DSN and we simply have to run the script (*nothing* else), the most straightforward way I could think of was writing a minimal Go script that accepts the DSN, a path to the file, and executes it as a whole.
  - However, the whole setup of the server, MySQL included, is built inside a network in Docker, so the script cannot reach MySQL without being inside it. Thus, I created a minimal Dockerfile on top of the `golang` image that simply builds and run the script.
- So if you can think of a way to run the `mysql` CLI directly (or some alternative), we could avoid having both the Go script and the Dockerfile wrapping it. Both are very minimal, but still.

As you'll see both in the Github action and in the Go script, there are many things hardcoded now. And that's why this is still a draft PR. So, in order to promote this to a real PR, I need to:
- [ ] ~~Generalize the action so that it can run with arbitrary versions and scripts. Probably having a reusable action that receives the window of versions for the migration, the migration script and the original DB dump against which the migrations will run.~~
- [x] Make the CI actually useful, since right now it always passes! We need to add a final step for `diff`ing both dumps and fail if the differences are not expected (`db_migrations` table, columns that are updated with `now()` during the migration and so on).

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-51095

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: .github/workflows/esrupgrade.yml
Comment (lieut-data) [inline .github/workflows/esrupgrade.yml:84]: Out of curiosity, can we just use the `mysql` tooling to ingest this file directly? What does the Go tooling offer?

File: scripts/esrupgrades/esr.5.37-7.8.mysql.up.sql
Comment (lieut-data) [inline scripts/esrupgrades/esr.5.37-7.8.mysql.up.sql:1]: For consistency, should we organize these into `mysql/` (and if necessary) `postgres/` folders to mirror the `db/` layout? (And minimize the chance of trying the wrong file? Not that it would run that far :P)

**Review by lieut-data — APPROVED** [review]
Looks great, @agarciamontoro! A few comments, but nothing blocking. 

To confirm, is this first pass just the "copy / paste" effort, and that we haven't done anything to examine how to "improve" this yet, e.g. merging some `ALTER TABLE` statements, etc.?

**Comment by lieut-data** [comment]
@toninis, wondering if you might be interested in looking at the new GH action here to validate the ESR upgrade script?

**Review by isacikgoz — COMMENTED** [review]
Looks good overall, a good start! Let's simplify this a bit and concentrate on the actual comparison part :)

File: .github/workflows/esrupgrade.yml
Comment (toninis) [inline .github/workflows/esrupgrade.yml:30]: You can use Github Actions  [containerised services](https://docs.github.com/en/actions/using-containerized-services/about-service-containers) in order to test this . 

You can check example [here](https://github.com/mattermost/mattermost-plugin-playbooks/blob/master/.github/workflows/ci.yaml#L292) 
I do not know if this suits your case but it is easier to maintain and you can spawn as many as you want without worrying if services are up etc

**Comment by toninis** [comment]
@isacikgoz Check [here](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontaineroptions) . Host network is disabled for Github Actions in general . I think it is for security reasons .

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:84]: Yes, the answer is what Ibrahim pointed out.

Your proposed command fails both locally and in CI with:
```
ERROR 1064 (42000) at line 25: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '' at line 17
```

Check this commit in the playground branch I have for testing: https://github.com/mattermost/mattermost-server/commit/6a96473216831720bc49459c02f4860c76834f23

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:8]: Yeah, this was copied from the other actions in the server. I don't have any preference regarding this, as long as it runs when we modify the files.

Is there a way to run the workflow manually?

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:30]: That's interesting, I'll take a look and see if it can be applied. In any case, I'd defer that to another PR, since this bit was copied from another action and it's been proved to work.

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:8]: Answering myself: https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow

However, running it manually doesn't seem to provide a good dev experience. We could have the same triggers we have now, but using the `paths` filter https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

File: .github/workflows/esrupgrade.yml
Comment (lieut-data) [reply .github/workflows/esrupgrade.yml:84]: Hmm, if we can't just do `mysql < script.sql`, how will our customers leverage this script?

Are we needing to do something like what's described in https://dba.stackexchange.com/questions/41336/how-to-create-a-mysql-stored-procedure-through-linux-terminal?

**Comment by lieut-data** [comment]
@agarciamontoro, for the `Reset DB` step, might we consider just dropping and recreating the database instead of trying to enumerate and drop tables with foreign key constraints removed? It might also be nice to use a different database for each of the two steps to guarantee the semantics.

**Comment by agarciamontoro** [comment]
@lieut-data, that's a good point. If I'm not mistaken (learning about Github actions as we speak), we could actually run both migrations in parallel in two different jobs so that both start with the exact same setup. Then reconcile both files in a third job that depends on them. I'll try that, thanks!

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:84]: That's a good point. I don't know how the customer that triggered all this effort actually runs their SQL commands. Let me ask, and depending on that we could test the `DELIMITER` stuff.

Comment (isacikgoz) [reply .github/workflows/esrupgrade.yml:92]: Interestingly I can run this command locally with our sql files :/ I don't know if I configured something else differently.

File: .github/workflows/esrupgrade.yml
Comment (isacikgoz) [reply .github/workflows/esrupgrade.yml:8]: My concern is why we should run this on every PR? What's going to change that often? We have *not so strict* rule to not touching older migrations. That means this will always run `v5.37` -> `v7.8` scripts. Isn't it?

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:92]: @isacikgoz, have you tried with the script in this PR? I've found the issue only with specific individual migrations, as well as with this script (understandable, since it contains such migrations).

Comment (isacikgoz) [reply .github/workflows/esrupgrade.yml:92]: Ah okay, that's not related to multiStatements 🤦 It's about the quotes, there should be a fix for that. Can you look into that if there is a fix? ref: https://stackoverflow.com/questions/75598918/stored-procedure-with-single-quotes

Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:113]: There are two hard problems at computer science and all three of them are wrapped in different quotes.

Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:113]: Now seriously, I'll try to see if this is the real cause and find a solution.

**Comment by agarciamontoro** [comment]
The diff is now down to a few rows in the `Roles` table. Check the artifact in https://github.com/mattermost/mattermost-server/actions/runs/4365145073/jobs/7633372640. I need to check the actual differences, since some of them seem to be just out of order permission names, and update timestamps.

@lieut-data, @isacikgoz, please take a look at the cleanup script. It's not finished, and we probably want to remove some stuff from it and add the actual migrations to the upgrade script instead. But I wanted to keep isolating the changes and listening to your thoughts. Thanks!

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:8]: Forgot to answer this before: we can simply run this on master (or on the release and cloud branches only), and on pull requests that modify files inside scripts/esrupgrades or .github/workflows/esr*. So something like (untested):
```yml
on:
  pull_request:
  push:
    branches:
      - cloud
      - release-*
    paths:
      - 'scripts/esrupgrades/*'
      - '.github/workflows/esr*'
```

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:56]: Yeah, probably. Do the actions have a general timeout?

File: scripts/esrupgrades/esr.5.37-7.8.mysql.cleanup.sql
Comment (lieut-data) [inline scripts/esrupgrades/esr.5.37-7.8.mysql.cleanup.sql:73]: nit: can we permalink to the GitHub branch where these last existed instead of embedding?

File: .github/workflows/esrupgrade.yml
Comment (isacikgoz) [reply .github/workflows/esrupgrade.yml:8]: Makes sense to me. Thanks!

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml]: So yes, the problem was indeed the lack of `DELIMITER` statements, not anything quote-related. The documentation is actually pretty clear on this topic: https://dev.mysql.com/doc/refman/8.0/en/stored-programs-defining.html
Please see https://github.com/mattermost/mattermost-server/pull/22448/commits/fac26e2eadeba019199866312f56d019c87f6cfd for a very comforting commit.

**Review by isacikgoz — COMMENTED** [review]
@agarciamontoro is this ready for the second pass?

File: scripts/esrupgrades/esr.5.37-7.8.mysql.cleanup.sql
Comment (agarciamontoro) [reply scripts/esrupgrades/esr.5.37-7.8.mysql.cleanup.sql]: Done! Improved docs in https://github.com/mattermost/mattermost-server/pull/22448/commits/88cad7915b0311f0f23f57573300721a30496d8a

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:63]: Done in https://github.com/mattermost/mattermost-server/pull/22448/commits/29d0355db5bd8b43331287512cd67515372acc39

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml]: Done in https://github.com/mattermost/mattermost-server/pull/22448/commits/01a47e96fbee8362860683ef0cfe28cb9d13acb9

File: .github/workflows/esrupgrade.yml
Comment (agarciamontoro) [reply .github/workflows/esrupgrade.yml:43]: Done.

**Comment by agarciamontoro** [comment]
@isacikgoz, it is now ready. I've changed my mind with respect to the original plan, which was to merge this PR only when CI was actually useful *and* when the action was general enough to allow for multiple ESR upgrades. The CI now fails when the diff is non-empty, so that requirement is fulfilled, but the action is still hard-coded to the 5.37 > 7.8 path.

However, I now think we should merge this PR as is (once you review and approve, I still want you to take a look at the cleanup script) and open a new one for modifying the script as needed, which is what we actually need as soon as possible. If time permits, we can even work in parallel in the generalization of the action. Thoughts?

**Comment by agarciamontoro** [comment]
With respect to the Roles differences I mentioned last time, those are gone now, thanks to a mildly complex procedure to standardize the table between migrations. So the diff is now empty! Both dumps are exactly the same. Please take a deep look at https://github.com/mattermost/mattermost-server/pull/22448/commits/88cad7915b0311f0f23f57573300721a30496d8a

**Review by lieut-data — APPROVED** [review]
Super slick work, @agarciamontoro!

**Review by isacikgoz — APPROVED** [review]
Brilliant work @agarciamontoro This looks good, thanks for addressing all of the requests along the way!

**Comment by agarciamontoro** [comment]
@lieut-data, @isacikgoz: is this safe to merge given the work for the monorepo?

**Comment by lieut-data** [comment]
@agarciamontoro, if possible, let's hold off merging this until the `mono-repo` changes. Everything should resolve either way, but it might help reduce churn on those larger changes right now.

**Comment by agarciamontoro** [comment]
Got it, thanks :)

**Comment by agarciamontoro** [comment]
Files are now in the correct place, CI is passing, all is good. It's been a while, but we can finally merge :tada:

**Pull request merged**
